src/webgpu/api/operation/adapter/info.spec.ts

```diff
@@ -39,3 +39,42 @@ g.test('adapter_info')
       `adapterInfo.device should be a normalized identifier. But it's '${adapterInfo.device}'`
     );
   });
+
+g.test('same_object')
+  .desc(
+    `
+GPUAdapter.info provides the same object each time it's accessed`
+  )
+  .fn(async t => {
+    const gpu = getGPU(t.rec);
+    const adapter = await gpu.requestAdapter();
+    assert(adapter !== null);
+
+    const adapterInfo1 = adapter.info;
+    const adapterInfo2 = adapter.info;
+    t.expect(adapterInfo1 === adapterInfo2);
+  });
+
+g.test('device_matches_adapter')
+  .desc(
+    `
+Test that GPUDevice.adapterInfo matches GPUAdapter.info`
+  )
+  .fn(async t => {
+    const gpu = getGPU(t.rec);
+    const adapter = await gpu.requestAdapter();
+    assert(adapter !== null);
+
+    const device = await t.requestDeviceTracked(adapter);
+    assert(device !== null);
+
+    assert(device.adapterInfo instanceof GPUAdapterInfo);
+    assert(adapter.info instanceof GPUAdapterInfo);
+
+    for (const k of Object.keys(GPUAdapterInfo.prototype)) {
+      t.expect(
+        objectEquals(device.adapterInfo[k], adapter.info[k]),
```

Comment:
Since browsers are allowed to make last-moment decisions about the contents of GPUAdapterInfo, I'd like to have a few variants on this test which access things in the following order:

- device.adapterInfo, adapter.info, device.adapterInfo[k1], adapter.info[k1], device.adapterInfo[k2], adapter.info[k2] (this is what you have now)
- same but flip adapter and device
- device.adapterInfo, device.adapterInfo[k1], device.adapterInfo[k2], adapter.info, adapter.info[k1], adapter.info[k2]
- same but flip adapter and device

(Each variant needs to get its own adapter and device, not reuse them.)

Just to be paranoid.

Reply:
@kainino0x `device.adapterInfo, device.adapterInfo[k1], device.adapterInfo[k2], adapter.info, adapter.info[k1], adapter.info[k2]` is a bit messy when coding it. Are we sure we want to be that paranoid for such as simple thing?

Reply:
It will be a little messy to write once but I think it is worth it. Just in case any browser ever does fingerprinting mitigations at this level, it could be very weird if they got this wrong.

To do that one I think you can basically do:

```
const deviceInfo: unknown[] = [];
for (const k of Object.keys(GPUAdapterInfo.prototype)) deviceInfo.push(device.adapterInfo[k]);
const adapterInfo: unknown[] = [];
for (const k of Object.keys(GPUAdapterInfo.prototype)) deviceInfo.push(adapter.info[k]);
//objectEquals(deviceInfo, adapterInfo)
```

Reply:
The potential complexity of this could be an argument in favor of having both return the same object... but I don't think it's really worth changing the spec further.